File: agent/parse.py
"""容错解析模型的 JSON 输出，迁自 lib/parse.ts。

模型偶尔在 JSON 外裹话或加代码块，这里宽松解析并归一化每个字段，免得一点抖动就
让整条流程崩或冒出垃圾。概念 name 从 JSON 来本就干净，不做符号清洗。
"""
import json
import re

LAYERS = ["宏观", "产业", "微观", "技术"]


def extract_json(raw: str) -> dict:
    """剥掉 ```json 代码块；否则切最外层的 {...}（检索版常在 JSON 外裹散文）。"""
    s = raw.strip()
    m = re.match(r"^```(?:json)?\s*([\s\S]*?)\s*```$", s, re.IGNORECASE)
    if m:
        s = m.group(1).strip()
    if not s.startswith("{"):
        a = s.find("{")
        b = s.rfind("}")
        if a != -1 and b > a:
            s = s[a : b + 1]
    return json.loads(s)


def pick_layer(raw: str) -> str:
    """layer 容错：只要包含四层之一就归一；都不含则默认 技术。"""
    for layer in LAYERS:
        if layer in raw:
            return layer
    return "技术"


def parse_structured(raw: str) -> dict:
    try:
        obj = extract_json(raw)
    except Exception:
        obj = {}  # 空/坏响应不崩流程，返回空结构, 交由 critic/评测发现
    if not isinstance(obj, dict):
        obj = {}  # 模型偶尔把整个响应返回成 JSON 数组/标量, 也按空处理
    concepts = []
    if isinstance(obj.get("concepts"), list):
        for c in obj["concepts"]:
            o = c or {}
            name = str(o.get("name", "")).strip()
            if not name:
                continue
            concepts.append(
                {
                    "name": name,
                    "layer": pick_layer(str(o.get("layer", ""))),
                    "definition": str(o.get("definition", "")).strip(),
                }
            )
    return {
        "summary": str(obj.get("summary", "")).strip(),
        "mechanism": str(obj.get("mechanism", "")).strip(),
        "uncertainty": str(obj.get("uncertainty", "")).strip(),
        "concepts": concepts,
    }


def parse_critique(raw: str) -> dict:
    """解析自检结果。解析失败时保守放行(passed=True)不阻塞流程；
    有 issues 却 passed=True 时以 issues 为准判不过(保守一致性)。"""
    try:
        obj = extract_json(raw)
    except Exception:
        return {"passed": True, "issues": []}
    if not isinstance(obj, dict):
        return {"passed": True, "issues": []}
    issues = []
    if isinstance(obj.get("issues"), list):
        for it in obj["issues"]:
            o = it if isinstance(it, dict) else {"detail": str(it)}  # 模型有时把 issue 写成纯字符串
            issues.append(
                {
                    "rule": str(o.get("rule", "")).strip(),
                    "severity": str(o.get("severity", "")).strip(),
                    "detail": str(o.get("detail", "")).strip(),
                    "fix_hint": str(o.get("fix_hint", "")).strip(),
                }
            )
    passed = bool(obj.get("passed", len(issues) == 0))
    if issues and passed:
        passed = False
    return {"passed": passed, "issues": issues}

File: lib/parse.ts
import { LAYERS, type GenerateArticleResult, type Layer } from './types';

// Shared, dependency-free parsing for the LLM's JSON output. Lives in its own
// module so both lib/llm.ts (single-shot) and lib/agent.ts (grounded) can reuse
// it without importing each other (avoids a circular dependency).

// The model is asked to return a JSON object. Parse it tolerantly and normalize
// every field, so a small wobble in the response can't crash the route or
// surface garbage. Concept names arrive clean from JSON — no symbol-stripping.
export function parseStructured(raw: string): GenerateArticleResult {
  const obj = extractJson(raw);
  const concepts = Array.isArray(obj.concepts)
    ? obj.concepts
        .map((c: unknown) => {
          const o = (c ?? {}) as Record<string, unknown>;
          return {
            name: String(o.name ?? '').trim(),
            layer: pickLayer(String(o.layer ?? '')),
            definition: String(o.definition ?? '').trim(),
          };
        })
        .filter((c) => c.name.length > 0)
    : [];
  return {
    summary: String(obj.summary ?? '').trim(),
    mechanism: String(obj.mechanism ?? '').trim(),
    uncertainty: String(obj.uncertainty ?? '').trim(),
    concepts,
  };
}

// Tolerant: strip a ```json fence if present, else slice the outermost { ... }.
// Grounding responses often wrap the JSON in prose, so the slice matters.
export function extractJson(raw: string): Record<string, unknown> {
  let s = raw.trim();
  const fence = s.match(/^```(?:json)?\s*([\s\S]*?)\s*```$/i);
  if (fence) s = fence[1].trim();
  if (!s.startsWith('{')) {
    const a = s.indexOf('{');
    const b = s.lastIndexOf('}');
    if (a !== -1 && b > a) s = s.slice(a, b + 1);
  }
  return JSON.parse(s) as Record<string, unknown>;
}

function pickLayer(raw: string): Layer {
  for (const l of LAYERS) {
    if (raw.includes(l)) return l;
  }
  console.warn(`[parse] unknown layer "${raw.trim()}", defaulting to 技术`);
  return '技术';
}
